### PR TITLE
Feat qasm to mpqp

### PR DESCRIPTION
**Pull request by JulienCalistoTD**

### CODE REVIEW

**Comment by Henri-ColibrITD** [comment]
write tests (we needed the random circ generation)

**Review by Henri-ColibrITD — APPROVED** [review]

File: mpqp/qasm/lexer_utils.py [hunk 1]
```diff
@@ -0,0 +1,176 @@
...
+
+one_parametrized_gate_qasm = {
+    "p": P,
+    "rx": Rx,
+    "ry": Ry,
+    "rz": Rz,
+}
+
+two_qubits_parametrized_gate_qasm = {
+    "cp": CRk,
```
Comment (hJaffaliColibritd) [inline mpqp/qasm/lexer_utils.py:139]: This means that each time we have a cp gate we will compute the tricky angle to have the right CRk corresponding?

File: mpqp/qasm/lexer_utils.py [hunk 2]
```diff
@@ -0,0 +1,176 @@
...
+}
+
+
+u_gate_qasm = {
+    "u": U,
```
[the remark this answers is not in the document]
Comment (JulienCalistoTD) [reply mpqp/qasm/lexer_utils.py:144]: done

File: mpqp/qasm/qasm_to_mpqp.py [hunk 3]
```diff
@@ -0,0 +1,349 @@
...
+
+def _Gate_two_qubits_parametrized(
+    circuit: QCircuit, gate_str: str, tokens: list[LexToken], idx: int
+) -> int:
+    if tokens[idx].type != 'LPAREN':
+        raise SyntaxError(f"Gate_one_parametrized: {idx} {tokens[idx]}")
+    idx += 1
+    parameter, idx = _eval_expr(tokens, idx)
+    if gate_str == 'cp':
+        parameter = np.log2(np.pi / parameter) + 1
```
Comment (hJaffaliColibritd) [inline mpqp/qasm/qasm_to_mpqp.py:228]: Looks like you handle here the parameter translation between the phase of `cp` and the order `k` of `CRk`.
However here your `k` will probably be a float, while `CRk` takes an integer I think.
As we discussed for the moment we can maybe round and put a todo for the future.

File: mpqp/qasm/qasm_to_mpqp.py [hunk 4]
```diff
@@ -0,0 +1,349 @@
...
+        raise SyntaxError(f"Gate_U: {idx} {tokens[idx]}")
+    idx += 1
+
+    theta, phi, lbda = 0, 0, 0
+    if gate_str == 'u1':
+        theta, idx = _eval_expr(tokens, idx)
+    elif gate_str == 'u2':
+        theta, idx = _eval_expr(tokens, idx)
+        phi, idx = _eval_expr(tokens, idx)
+    elif gate_str == 'u3' or gate_str == 'u':
```
Comment (hJaffaliColibritd) [inline mpqp/qasm/qasm_to_mpqp.py:338]:
```suggestion
    elif gate_str == 'u3' or gate_str == 'u' or gate_str == 'U':
```

File: tests/qasm/test_qasm_to_mpqp.py [hunk 5]
```diff
@@ -0,0 +1,200 @@
...
+                H(0),
+                CNOT(0, 1),
+                BasisMeasure([0], [2]),
+                BasisMeasure([1], [1]),
+                BasisMeasure([2], [0]),
+            ],
+        ),
+    ],
+)
+def test_qasm2_to_mpqp(qasm_code: str, gate_names: list[str]):
```
Comment (hJaffaliColibritd) [inline tests/qasm/test_qasm_to_mpqp.py:168]: To follow the discussion we had, maybe add a test with `cp`.

File: tests/qasm/test_qasm_to_mpqp.py [hunk 6]
```diff
@@ -0,0 +1,200 @@
...
+    ],
+)
+def test_invalid_qasm_code(qasm_code: str):
+    try:
+        qasm2_parse(qasm_code)
+    except SyntaxError:
+        pass
+
+
+def test_random_qasm_code():
```
Comment (hJaffaliColibritd) [inline tests/qasm/test_qasm_to_mpqp.py:194]: Here we have a test for pure import.
Maybe we should check the result of execution also ?

**Review by hJaffaliColibritd — CHANGES_REQUESTED** [review]

File: mpqp/qasm/qasm_to_mpqp.py (hunk 3 above)
Comment (JulienCalistoTD) [reply mpqp/qasm/qasm_to_mpqp.py:228]: yes, CRk can take float

File: mpqp/qasm/qasm_to_mpqp.py (hunk 4 above)
Comment (JulienCalistoTD) [reply mpqp/qasm/qasm_to_mpqp.py:338]: done

File: tests/qasm/test_qasm_to_mpqp.py (hunk 5 above)
Comment (JulienCalistoTD) [reply tests/qasm/test_qasm_to_mpqp.py:168]: done

**Pull request merged**